File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/geo/GeoJson.java
package org.maplibre.nativeffi.geo;

import java.util.List;
import java.util.Objects;

/** Immutable GeoJSON descriptor tree. */
public sealed interface GeoJson
    permits GeoJson.GeometryValue, GeoJson.FeatureValue, GeoJson.FeatureCollection {
  static GeometryValue geometry(Geometry geometry) {
    return new GeometryValue(geometry);
  }

  static FeatureValue feature(Feature feature) {
    return new FeatureValue(feature);
  }

  static FeatureCollection featureCollection(List<Feature> features) {
    return new FeatureCollection(features);
  }

  record GeometryValue(Geometry geometry) implements GeoJson {
    public GeometryValue {
      Objects.requireNonNull(geometry, "geometry");
    }
  }

  record FeatureValue(Feature feature) implements GeoJson {
    public FeatureValue {
      Objects.requireNonNull(feature, "feature");
    }
  }

  record FeatureCollection(List<Feature> features) implements GeoJson {
    public FeatureCollection {
      features = List.copyOf(features);
    }
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/geo/GeoJson.kt
package org.maplibre.nativeffi.geo

/** Immutable GeoJSON descriptor tree. */
public sealed interface GeoJson {
  public data class GeometryValue(public val geometry: Geometry) : GeoJson

  public data class FeatureValue(public val feature: Feature) : GeoJson

  public class FeatureCollection(features: List<Feature>) : GeoJson {
    public val features: List<Feature> = features.toList()

    override fun equals(other: Any?): Boolean =
      other is FeatureCollection && features == other.features

    override fun hashCode(): Int = features.hashCode()

    override fun toString(): String = "FeatureCollection(features=$features)"
  }
}
